### PR TITLE
Implement RobustArmijo and refactor Armijo Line Search

### PR DESCRIPTION
**Pull request by zfergus**
* Implement the numerically robust Armijo line search of Longva et al. [2023]

TODO:
* [x] Re-enable `use_grad_norm`
* [x] Only compute `new_grad` as needed

### CODE REVIEW

File: src/polysolve/nonlinear/line_search/RobustArmijo.cpp [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        const double old_energy,
+        const TVector &old_grad,
+        const double new_energy,
+        const TVector &new_grad,
+        const double step_size) const
+    {
+        if (new_energy <= old_energy + step_size * this->armijo_criteria) // Try Armijo first
+            return true;
+
+        if (std::abs(new_energy - old_energy) <= 0.1 * std::abs(old_energy))
```
Comment (teseoch) [inline src/polysolve/nonlinear/line_search/RobustArmijo.cpp:25]: 0.1 json input?

File: src/polysolve/nonlinear/line_search/RobustArmijo.cpp [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        {
+            // TODO: Compute new_grad here as needed
+            // objFunc.gradient(new_x, new_grad);
+            assert(old_grad.size() == new_grad.size());
+
+            const double deltaE_approx = step_size / 2 * delta_x.dot(new_grad + old_grad);
```
Comment (teseoch) [inline src/polysolve/nonlinear/line_search/RobustArmijo.cpp:31]: /2 should be based on step_ratio?

File: src/polysolve/nonlinear/line_search/Backtracking.cpp [hunk 3]
```diff
@@ -20,58 +18,64 @@ namespace polysolve::nonlinear::line_search
...
             }
-            else
-                cur_energy = objFunc.value(new_x);
 
-            is_step_valid = objFunc.is_step_valid(x, new_x);
+            const double new_energy = objFunc.value(new_x);
 
-            m_logger.trace("ls it: {} delta: {} invalid: {} ", this->cur_iter, (cur_energy - old_energy), !is_step_valid);
+            TVector new_grad; // TODO: only compute gradient if needed
+            objFunc.gradient(new_x, new_grad);
```
[the remark this answers is not in the document]
Comment (zfergus) [reply src/polysolve/nonlinear/line_search/Backtracking.cpp:52]: The RobustArmijo needs it, but it should be computed as needed in the criteria function for example.

File: src/polysolve/nonlinear/line_search/Backtracking.cpp [hunk 4]
```diff
@@ -20,58 +18,64 @@ namespace polysolve::nonlinear::line_search
...
+
+            if (criteria(delta_x, old_energy, old_grad, new_energy, new_grad, step_size))
             {
-                break;
+                break; // found a good step size
             }
-            this->cur_iter++;
         }
 
         return step_size;
```
[the remark this answers is not in the document]
Comment (zfergus) [reply src/polysolve/nonlinear/line_search/Backtracking.cpp:67]: In the for loop's increment statement. I thought it was cleaner that way we can just use continue anywhere and the stepsize and cur_iter are updated automatically.

File: src/polysolve/nonlinear/line_search/RobustArmijo.cpp (hunk 1 above)
Comment (zfergus) [reply src/polysolve/nonlinear/line_search/RobustArmijo.cpp:25]: yeah, I'll do that

**Review by teseoch — APPROVED** [review]

**Pull request merged**